pkg/kv/kvserver/loqrecovery/plan.go

```diff
@@ -119,22 +112,109 @@ type ReplicaUpdateReport struct {
 // An error is returned in case of unrecoverable error in the collected data
 // that prevents creation of any sane plan or correctable user error.
 func PlanReplicas(
-	ctx context.Context, nodes []loqrecoverypb.NodeReplicaInfo, deadStores []roachpb.StoreID,
+	ctx context.Context,
+	nodes []loqrecoverypb.NodeReplicaInfo,
+	descriptors []roachpb.RangeDescriptor,
+	deadStoreIDs []roachpb.StoreID,
+	deadNodeIDs []roachpb.NodeID,
 ) (loqrecoverypb.ReplicaUpdatePlan, PlanningReport, error) {
-	var report PlanningReport
-	updatedLocations := make(updatedLocationsMap)
 	var replicas []loqrecoverypb.ReplicaInfo
 	for _, node := range nodes {
 		replicas = append(replicas, node.Replicas...)
 	}
-	availableStoreIDs, missingStores, err := validateReplicaSets(replicas, deadStores)
+	availableStoreIDs, deadNodes, err := validateReplicaSets(replicas, deadStoreIDs, deadNodeIDs)
 	if err != nil {
 		return loqrecoverypb.ReplicaUpdatePlan{}, PlanningReport{}, err
 	}
-	report.PresentStores = storeSliceFromSet(availableStoreIDs)
-	report.MissingStores = storeSliceFromSet(missingStores)
-
 	replicasByRangeID := groupReplicasByRangeID(replicas)
+
+	var report PlanningReport
+	report.MissingNodes = deadNodes.asSortedSlice()
+	report.PresentStores = availableStoreIDs.storeSliceFromSet()
+	report.TotalReplicas = len(replicas)
+
+	if len(descriptors) > 0 {
+		report.PlannedUpdates, report.Problems, err = planReplicasWithMeta(ctx, descriptors, replicasByRangeID, availableStoreIDs)
+	} else {
+		report.PlannedUpdates, report.Problems, err = planReplicasWithoutMeta(ctx, replicasByRangeID, availableStoreIDs)
+	}
+	if err != nil {
+		return loqrecoverypb.ReplicaUpdatePlan{}, PlanningReport{}, err
+	}
+
+	updates := make([]loqrecoverypb.ReplicaUpdate, len(report.PlannedUpdates))
+	updatedLocations := make(locationsMap)
+	for i, u := range report.PlannedUpdates {
+		updates[i] = u.asReplicaUpdate()
+		updatedLocations.add(u.NewReplica.NodeID, u.NewReplica.StoreID)
+		report.DiscardedNonSurvivors += u.DiscardedReplicasCount
+	}
+	report.UpdatedNodes = updatedLocations.asSortedSlice()
+
+	var decommissionedNodeIDs []roachpb.NodeID // this is all wrong!
+	for id := range updatedLocations {
+		decommissionedNodeIDs = append(decommissionedNodeIDs, id)
+	}
```

Comment:
I know this is from a previous PR, but we should call this `DecommissionNodeIDs` rather than `Decommissioned`, to clarify that they will be marked as decommissioned but they aren't yet.